### PR TITLE
Tpetra: Refactor CrsMatrix pack/unpack procedures to use PackTraits

### PR DESCRIPTION
**Pull request by tjfulle**
@trilinos/tpetra, @mhoemmen

Comments
--------
This commit is a combination of several commits that address a several
issues: #797, #798, #800, #802

A summary of changes are as follows:

- Refactor CrsMatrix pack/unpack procedures to use PackTraits (for both static
  and dynamic profile matrices)
- Refactor packCrsMatrix to pack (optional) PIDs
- Remove exists packAndPrepareWithOwningPIDs and instead use the aforementioned
  packCrsMatrix procedure.
- Modify PackTraits run on threads by removing calls to TEUCHOS_TEST_FOR_EXCEPTION
  and decorating device code with KOKKOS_INLINE_FUNCTION
- Ditto for Stokhos' specialization of PackTraits

Build/Test Cases Summary
Enabled Packages: TpetraCore
Disabled Packages: PyTrilinos,Claps,TriKota
Enabled all Forward Packages
0) MPI_RELEASE_DEBUG_SHARED_PT => passed:
passed=1483,notpassed=0 (79.62 min)

### CODE REVIEW

**Comment by tjfulle** [comment]
This is the conclusion of #1535

**Comment by tjfulle** [comment]
@etphipp , one Stokhos file was modified as part of this PR

**Comment by tjfulle** [comment]
Copy of email I sent @mhoemmen :

Not sure if you get GitHub notifications, but I just reopened a PR for the PackTraits and friends work.  It is a single commit that:

- Refactors the dynamic and static CrsMatrix pack/unpack to us PackTraits
- Refactors packCrsMatrix to pack optional PIDs
- Refactors parts of unpackCrsMatrix to unpack optional PIDs
- Removes Import_Util::packAndPrepareWithOwningPIDs in favor of Details::packCrsMatrixWithOwningPIDs
- Modifies PackTraits to be device friendly
- Modifies Stokhos' specialization of PackTraits to be consistent with Tpetra::Details::PackTraits
- Modifies tests for new interfaces

The single commit was bigger than I like.  I had originally made many independent commits, but squished them all together since the downstream ROL example (that broke) demonstrated that most of the intermediate commits had a bug (that didn't manifest in other tests).  Rather than pushing broken commits, I just squished them all so the final package is a single working commit

**Comment by mhoemmen** [comment]
btw the conflict is just because packAndPrepareWithOwningPIDs got absorbed into the new pack stuff.  @DrBooom assures me that this won't conflict with his work on communication optimization for sparse matrix-matrix multiply, assuming that the pack format hasn't changed.

**Comment by tjfulle** [comment]
The pack format is identical

**Comment by mhoemmen** [comment]
@tjfulle Argh, I should have looked more closely at this -- Victor tried it out and it looks like it doesn't build with complex enabled.  I can work on the fix if you don't have time.

**Comment by tjfulle** [comment]
@mhoemmen , this pull request hasn't been merged.  Unless Victor is using my branch, I don't see how it is related?

**Comment by mhoemmen** [comment]
@tjfulle Victor is trying out your branch :-)  Don't worry, I can fix it pretty easily :-)

**Comment by mhoemmen** [comment]
@tjfulle Argh, I don't know what I just did but I think I fixed the merge conflict?  Let me try again.

**Comment by mhoemmen** [comment]
@tjfulle I've got a patch ready to fix complex builds.  GitHub won't let me upload it here for some reason, even when I rename its extension to .txt (it says "Something went really wrong, and we can't process that file"), so I'll just e-mail it to you.  I think you could just squash it into this one commit, in order to preserve atomicity; you shouldn't have to redo the merge with develop commit.

**Comment by tjfulle** [comment]
@mhoemmen, Okay.  I'm getting close with the latest ROL/Panzer failure.  When I get this resolved, I'll merge the file you send me and see how the tests look.

**Comment by mhoemmen** [comment]
> Okay. I'm getting close with the latest ROL/Panzer failure. When I get this resolved, I'll merge the file you send me and see how the tests look.

Thanks and no worries! :-D

**Comment by mhoemmen** [comment]
Tpetra tests pass with CUDA with this patch, btw!  (except for a gemv test, but that's because the error bounds are too tight for complex -- I should be able to fix that separately)

**Comment by tjfulle** [comment]
@mhoemmen , I've got errors in a ROL example.  It passes with `OMP_NUM_THREADS=1` but fails for `OMP_NUM_THREADS>1`.  I should have a fix soon then this might be done!

**Comment by mhoemmen** [comment]
I've been working on this -- got it to build but Tpetra tests not passing yet -- should finish tomorrow sometime.  The issue was that multiple threads were sharing the same scratch space.  I'm working on a patch to fix this.

**Comment by tjfulle** [comment]
@mhoemmen, I've applied your patches and modified allocation of scratch arrays in `unpackAndCombine` to be separate for each thread.  With the changes, the previously failing Panzer and ROL tests pass (with `OMP_NUM_THREADS=4`).  My fix is temporary, and only works on CPU+OpenMP.  If you send me a patch with `Kokkos::MemoryPool`, I can apply it and run all the tests again.  Or, I can push what I have and you can put it in directly, if you prefer.

**Comment by mhoemmen** [comment]
@tjfulle wrote:
> @mhoemmen, I've applied your patches and modified allocation of scratch arrays in unpackAndCombine to be separate for each thread. With the changes, the previously failing Panzer and ROL tests pass (with OMP_NUM_THREADS=4). My fix is temporary, and only works on CPU+OpenMP. 

As long as you aren't reallocating and deallocating scratch on each loop iteration, I would say this is fine for now.  Just run everything on host, by using `Kokkos::View<int*, device_type>::HostMirror::execution_space` as the execution space in the `Kokkos::RangePolicy` for the parallel loop.

> If you send me a patch with Kokkos::MemoryPool, I can apply it and run all the tests again. Or, I can push what I have and you can put it in directly, if you prefer.

The latter, please.  I will need to think harder about MemoryPool; I think it's overkill for this application and may have some overhead as a result.

**Comment by mhoemmen** [comment]
@tjfulle Just to summarize our out-of-band conversation: Please leave this as a PR for now.  We should do some performance testing before pushing it.  Victor can help with that.

**Comment by tjfulle** [comment]
@mhoemmen.  The latest push merges your fixes for complex and fixes cases for `OMP_NUM_THREADS > 1`.

**Comment by tjfulle** [comment]
@mhoemmen, I just rebased a single commit on top of the (current) trilinos/develop HEAD.  Standard checkin tests pass, standard checkin tests with `OpenMP` node type and `OMP_NUM_THREADS=2` pass, Tpetra tests pass with `CUDA` node type (on `ride`) with `OMP_NUM_THREADS=2`.

**Comment by mhoemmen** [comment]
Thanks @tjfulle !  I just got a CUDA 8 build set up after a CUDA driver upgrade; I'll test downstream of Tpetra.  If it works, I'll merge it!

**Comment by mhoemmen** [comment]
Some weird thing happened this morning with the Kokkos CUDA build -- nothing to do with Tpetra -- I am trying to figure it out.  Once I do that, I'll test this PR.

**Comment by mhoemmen** [comment]
I actually found (and fixed) some unrelated bugs by testing CUDA downstream.  Ifpack2 is fine now.  I'm having trouble linking MueLu in my debug CUDA build, even with `-Os`.  I'm just going to run Ross' remote check-in test script and push the thing; I don't want you to have to keep waiting on the various CUDA build issues that have come up and keep coming up.

**Comment by mhoemmen** [comment]
Hi @tjfulle !  I pushed this commit via the check-in test script.  Thanks for working on this!
